pkg/cloudprovider/providers/vsphere/vsphere.go

```diff
@@ -1374,3 +1647,95 @@ func makeDirectoryInDatastore(c *govmomi.Client, dc *object.Datacenter, path str
 
 	return err
 }
+
+// Gets the folder path from the VM Disk path.
+func getFolderFromVMDiskPath(vmDiskPath string) string {
+	i := strings.Index(vmDiskPath, "/")
+	folderPath := ""
+	if i > -1 {
+		folderPath = vmDiskPath[:i+1]
+	}
+	return folderPath
+}
+
+// Delete the VM.
+// An attempt is made to delete the VM for DeleteVMAttempts times before it backs off.
+func deleteVM(ctx context.Context, vm *object.VirtualMachine) error {
+	var destroyTask *object.Task
+	var err error
+	for i := 0; i < DeleteVMAttempts; i++ {
+		err = nil
+		// Delete the dummy VM
+		destroyTask, err = vm.Destroy(ctx)
+		if err != nil {
+			continue
+		} else {
+			err = destroyTask.Wait(ctx)
+			if err != nil {
+				continue
+			} else {
+				break
+			}
+		}
+	}
+	return err
+}
+
+// Get the folder
+func getFolder(ctx context.Context, c *govmomi.Client, datacenterName string, folderName string) (*object.Folder, error) {
+	f := find.NewFinder(c.Client, true)
+
+	// Fetch and set data center
+	dc, err := f.Datacenter(ctx, datacenterName)
+	if err != nil {
+		return nil, err
+	}
+	f.SetDatacenter(dc)
+
+	folderName = strings.TrimSuffix(folderName, "/")
+	dcFolders, err := dc.Folders(ctx)
+	vmFolders, _ := dcFolders.VmFolder.Children(ctx)
+
+	var vmFolderRefs []types.ManagedObjectReference
+	for _, vmFolder := range vmFolders {
+		vmFolderRefs = append(vmFolderRefs, vmFolder.Reference())
+	}
+
+	// Get only references of type folder.
+	var folderRefs []types.ManagedObjectReference
+	for _, vmFolder := range vmFolderRefs {
+		if vmFolder.Type == "Folder" {
+			folderRefs = append(folderRefs, vmFolder)
+		}
+	}
+
+	// Find the specific folder reference matching the folder name.
+	var resultFolder *object.Folder
+	pc := property.DefaultCollector(c.Client)
+	for _, folderRef := range folderRefs {
+		var refs []types.ManagedObjectReference
+		var folderMorefs []mo.Folder
+		refs = append(refs, folderRef)
+		err = pc.Retrieve(ctx, refs, []string{"name"}, &folderMorefs)
+		for _, fref := range folderMorefs {
+			if fref.Name == folderName {
+				resultFolder = object.NewFolder(c.Client, folderRef)
+			}
+		}
+	}
+
+	return resultFolder, nil
+}
+
+// Move the Virtual Disk to Kubevols directory.
+func moveVirtualDisk(ctx context.Context, c *govmomi.Client, dc *object.Datacenter, srcPath string, destPath string) error {
+	// Create a virtual disk manager
+	virtualDiskManager := object.NewVirtualDiskManager(c.Client)
+	// Move the Virtual disk from VM folder to the kubevols directory.
+	task, err := virtualDiskManager.MoveVirtualDisk(ctx, srcPath, dc, destPath, dc, true)
```

Comment:
Does this disk move retain VMDK property like (thin, ezt)? I think we will be better off creating VMDK at right place rather than moving it later.